File: LeetCode/79_word_search/DOS.py
'''
- still not used to backtracking.
'''
from typing import List


class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        nrow = len(board)
        ncol = len(board[0])
        up, down, left, right = (0,1), (0,-1), (-1,0), (1,0)
    
        def backtrack(i,j,runes):
            # basecase
            if not runes:
                return True
            if i < 0 or i == nrow or j < 0 or j == ncol or board[i][j] != runes[0]:
                return False

            # recursion
            result = False
            board[i][j] = '!' # visiting i,j
            for dx, dy in [up, down, left, right]:
                result = backtrack(i+dx, j+dy, runes[1:])
                if result:
                    break
            board[i][j] = runes[0] # complete visit; restore value
            return result
            
        for i in range(nrow):
            for j in range(ncol):
                if backtrack(i,j,word):
                    return True
        return False
            

File: LeetCode/22_Generate_Parentheses/DOS.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function (n) {
	let result = [];
	let count = n * 2;

	const backtrack = (S="", left=0, right=0) => {
		if (S.length === count) {
			result.push(S)
			return
		}
		if (left < n) {
			S += "(";
			backtrack(S, left+1, right);
			S = S.slice(0,-1);
		}
		if (right < left) {
			S += ")";
			backtrack(S, left, right+1);
			S = S.slice(0,-1);
		}
	}
	backtrack();
	return result
};

console.log(generateParenthesis(3));
